clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -632,9 +644,68 @@ public void beginTransaction() throws ProducerFencedException {
      */
     public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
                                          String consumerGroupId) throws ProducerFencedException {
+        if (!cachedGroupMetadata.groupId().equals(consumerGroupId)) {
+            // Generally this logic should only be triggered once during first call.
+            log.warn("Cached consumer groupId changed from {} to {}. If the old group id is not empty, this indicates an abuse of this API",
+                cachedGroupMetadata.groupId(), consumerGroupId);
+            cachedGroupMetadata = new ConsumerGroupMetadata(consumerGroupId,
+                JoinGroupRequest.UNKNOWN_GENERATION_ID, JoinGroupRequest.UNKNOWN_MEMBER_ID, Optional.empty());
+        }
+        sendOffsetsToTransactionInternal(offsets, cachedGroupMetadata, false);
+    }
+
+    /**
+     * Sends a list of specified offsets to the consumer group coordinator, and also marks
+     * those offsets as part of the current transaction. These offsets will be considered
+     * committed only if the transaction is committed successfully. The committed offset should
+     * be the next message your application will consume, i.e. lastProcessedMessageOffset + 1.
+     * <p>
+     * This method should be used when you need to batch consumed and produced messages
+     * together, typically in a consume-transform-produce pattern. Thus, the specified
+     * {@code groupMetadata} should be extracted from the used {@link KafkaConsumer consumer} via
+     * {@link KafkaConsumer#groupMetadata()} to leverage consumer group metadata for proper fencing.
+     * Note, that the consumer should have {@code enable.auto.commit=false} and should
+     * also not commit offsets manually (via {@link KafkaConsumer#commitSync(Map) sync} or
+     * {@link KafkaConsumer#commitAsync(Map, OffsetCommitCallback) async} commits).
+     *
+     * This API won't deprecate the existing {@link KafkaProducer#sendOffsetsToTransaction(Map, String) sendOffsets} API as standalone
```

Comment:
I don't think this comment should be in the javadoc. If it did deprecate the other API, we would just mark the other as deprecated.